src/libretro_ui.cpp

```diff
@@ -0,0 +1,739 @@
+	if (axis < -JOYSTICK_AXIS_SENSIBILITY) {
+		keys[Input::Keys::JOY_AXIS_Y_UP] = true;
+		keys[Input::Keys::JOY_AXIS_Y_DOWN] = false;
+	} else if (axis > JOYSTICK_AXIS_SENSIBILITY) {
+		keys[Input::Keys::JOY_AXIS_Y_UP] = false;
+		keys[Input::Keys::JOY_AXIS_Y_DOWN] = true;
+	} else {
+		keys[Input::Keys::JOY_AXIS_Y_UP] = false;
+		keys[Input::Keys::JOY_AXIS_Y_DOWN] = false;
+	}
+#	endif
+#	endif
+}
+
+bool LibretroUi::IsFullscreen() {
+	// Handled by libretro
+	return false;
+}
+
+uint32_t LibretroUi::GetTicks() const {
+	return (uint32_t)(time_in_microseconds/1000);
+}
+
+void LibretroUi::Sleep(uint32_t) {
+	// Sleep is not needed libretro will ensure 60 fps
+}
+
+retro_video_refresh_t LibretroUi::UpdateWindow = nullptr;
+void LibretroUi::SetRetroVideoCallback(retro_video_refresh_t cb) {
+	UpdateWindow = cb;
+}
+
+retro_input_state_t LibretroUi::CheckInputState = nullptr;
+void LibretroUi::SetRetroInputStateCallback(retro_input_state_t cb) {
+	CheckInputState = cb;
+}
+
+void LibretroUi::UpdateKeyboardCallback(bool down, unsigned keycode) {
+	if ((keyboard_retropad_state & PadInputState_Keyboard) == PadInputState_Keyboard) {
+		keys[RetroKey2InputKey(keycode)] = down;
+	}
+}
+
+void LibretroUi::UpdateVariables() {
+	static const char* none = "None (See Core Options)";
+
+	static struct retro_variable debug = { "Debug Mode", nullptr };
+	static struct retro_variable keyboard = { "Keyboard", nullptr };
+	static struct retro_variable variables[] = {
+		{ "RetroPad X", nullptr },
+		{ "RetroPad L", nullptr },
+		{ "RetroPad R", nullptr },
+		{ "RetroPad L2", nullptr },
+		{ "RetroPad R2", nullptr },
+		{ "RetroPad L3", nullptr },
+		{ "RetroPad R3", nullptr }
+	};
+
+	static const char buttons[][24] = {
+		"0", "1", "2", "3", "4",
+		"5", "6", "7", "8", "9",
+		"+", "-", "*", "/", ".",
+		"Open Debug Menu", "Walk through walls"
+	};
+
+	LibretroUi::environ_cb(RETRO_ENVIRONMENT_GET_VARIABLE, &debug);
+	Player::debug_flag = strcmp(debug.value, "Enabled") == 0;
+
+	LibretroUi::environ_cb(RETRO_ENVIRONMENT_GET_VARIABLE, &keyboard);
+	if (strcmp(keyboard.value, "Use Both") == 0) {
+		keyboard_retropad_state = PadInputState_Keyboard | PadInputState_RetroPad;
+	} else if (strcmp(keyboard.value, "Only Keyboard") == 0) {
+		keyboard_retropad_state = PadInputState_Keyboard;
+	} else if (strcmp(keyboard.value, "Only RetroPad") == 0) {
+		keyboard_retropad_state = PadInputState_RetroPad;
+	}
+
+	// Button remapping from settings
+	for (int i = 0; i < 7; ++i) {
+		LibretroUi::environ_cb(RETRO_ENVIRONMENT_GET_VARIABLE, &variables[i]);
+		button_remapper[i] = Input::Keys::NONE;
+		for (int j = 0; j < 17; ++j) {
+			if (strcmp(variables[i].value, buttons[j]) == 0) {
+				button_remapper[i] = Input::Keys::JOY_10 + j;
+				break;
+			}
+		}
+		if (button_remapper[i] == Input::Keys::NONE) {
+			variables[i].value = none;
+		}
+	}
+
+	struct retro_input_descriptor desc[] = {
```

Comment:
Guess you will miss this during the review, so I highlight this quirk here:
libretro requires that only the string-pointer (the last struct element) stays valid but as you can see this struct is on the stack and destroyed but as you know in C the struct doesn't contain the string but a pointer to the string and the strings are constant/known during compile time, so they are always pointing to the same valid storage in the data segment. :D